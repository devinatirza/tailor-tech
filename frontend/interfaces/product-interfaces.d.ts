export interface IProduct {
    Id: number;
    Name: string;
    Desc: string;
    Price: number;
    ImgUrl: string;
  }