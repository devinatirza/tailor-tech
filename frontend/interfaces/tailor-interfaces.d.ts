export interface ITailor{
    ID: number
    Name:  string
    Email: string
    Address: string
    ImgUrl: string
    Rating: float32
  }