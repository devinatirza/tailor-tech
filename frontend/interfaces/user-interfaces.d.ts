export interface IUser {
    id: number
    name: string
    email: string
    phoneNumber: string
    address: string
    points: number
}