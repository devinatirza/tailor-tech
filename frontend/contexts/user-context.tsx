import { createContext, useContext, useState } from "react";
import { IUser } from "../interfaces/user-interfaces";
import axios from "axios";

const userContext = createContext({} as any)

type ContentLayout = {
    children: JSX.Element
}

export function UserProvider({children}: ContentLayout){
    const [user, setUser] = useState<IUser|null>(null)

    async function login(email: string, pass: string){
        try{
            const response = await axios.post("http://localhost:8000/login", {
                email: email,
                pass: pass
            })

            if(response.status === 200){
                setUser(response.data)
                return ''
            }
            else{
                return 'An error occured.'
            }
        } catch(error){
            return error.response?.data.error
        }
    }

    const data = {user, login}

    return <userContext.Provider value={data}>{children}</userContext.Provider>
}

export function useUser(){
    return useContext(userContext)
}