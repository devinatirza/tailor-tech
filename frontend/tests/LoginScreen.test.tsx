import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { render, fireEvent, waitFor } from '@testing-library/react-native';
import LoginScreen from '../pages/Login';
import { IStackScreenProps } from '../src/library/StackScreenProps';
import { useNavigation } from '@react-navigation/native';
import { useUser } from '../contexts/user-context';
import axios from 'axios';
import { IUser } from '../interfaces/user-interfaces';
import { TextInput, TouchableOpacity, Text } from 'react-native';

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    navigate: jest.fn(),
  }),
}));

jest.mock('../contexts/user-context', () => ({
  useUser: jest.fn(),
}));

jest.mock('axios');

describe('LoginScreen', () => {
  const mockNavigation = { navigate: jest.fn() };
  const mockUserContext = {
    login: jest.fn() as jest.MockedFunction<any>,
    updateUser: jest.fn() as jest.MockedFunction<any>,
    user: null as IUser | null,
  };

  beforeEach(() => {
    (useUser as jest.Mock).mockReturnValue(mockUserContext);
    jest.clearAllMocks();
  });

  const mockProps: IStackScreenProps = {
    navigation: mockNavigation,
    route: {},
    nameProp: 'TestName',
  };

  it('renders login screen correctly (snapshot)', () => {
    const tree = renderer.create(<LoginScreen {...mockProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles successful login', async () => {
    mockUserContext.login.mockResolvedValue('');

    const component = renderer.create(<LoginScreen {...mockProps} />);
    const instance = component.root;

    const emailInput = instance.findAllByType(TextInput)[0];
    const passwordInput = instance.findAllByType(TextInput)[1];
    const loginButton = instance.findAllByType(TouchableOpacity).find(node => node.findByType(Text).props.children === 'Login');

    // if (emailInput && passwordInput && loginButton && loginButton.props && typeof loginButton.props.onPress === 'function') {
      await act(async () => {
        emailInput?.props.onChangeText('test@example.com');
        passwordInput?.props.onChangeText('password123');
      });

      await act(async () => {
        // await Promise.resolve();
        loginButton?.props.onPress();
      });

      // console.log(loginButton[0].findByType(Text).props.children)


      expect(mockUserContext.login).toHaveBeenCalledWith('test@example.com', 'password123');
      expect(mockNavigation.navigate).toHaveBeenCalledWith('TailorTech');
      expect(component.toJSON()).toMatchSnapshot();
    // } else {
    //   throw new Error('Login button or its onPress method not found');
    // }
  });

  it('handles failed login', async () => {
    mockUserContext.login.mockResolvedValue('Invalid credentials');

    const component = renderer.create(<LoginScreen {...mockProps} />);
    const instance = component.root;

    const emailInput = instance.findAllByType(TextInput)[0];
    const passwordInput = instance.findAllByType(TextInput)[1];
    const loginButton = instance.findAllByType(TouchableOpacity).find(node => node.findByType(Text).props.children === 'Login');

    // if (loginButton && loginButton.props && typeof loginButton.props.onPress === 'function') {
      await act(async () => {
        emailInput.props.onChangeText('test@example.com');
        passwordInput.props.onChangeText('wrongpassword');
        loginButton?.props.onPress();
      });

      await act(async () => {
        await Promise.resolve();
      });

      expect(mockUserContext.login).toHaveBeenCalledWith('', '');
      expect(mockNavigation.navigate).not.toHaveBeenCalledWith('TailorTech');
      expect(instance.findByProps({ children: 'Invalid credentials' })).toBeTruthy();
      expect(component.toJSON()).toMatchSnapshot();
    // } else {
    //   throw new Error('Login button or its onPress method not found');
    // }
  });

  it('navigates to Register screen', async () => {
    const component = renderer.create(<LoginScreen {...mockProps} />);
    const instance = component.root;

    const registerButton = instance.findAllByType(TouchableOpacity).find(node => node.findByType(Text).props.children === "Didn't have an account?");

    await act(async () => {
      registerButton?.props.onPress();
    });

    await act(async () => {
      await Promise.resolve();
    });

    expect(mockNavigation.navigate).toHaveBeenCalledWith('Register');
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('validates user on mount', async () => {
    const mockValidatedUser = { id: 1, name: 'John Doe', email: 'john.doe@example.com' };
    (axios.get as jest.MockedFunction<any>).mockResolvedValue({ data: mockValidatedUser });

    await act(async () => {
      renderer.create(<LoginScreen {...mockProps} />);
    });

    expect(mockUserContext.updateUser).toHaveBeenCalledWith(mockValidatedUser);
    if (mockUserContext.user) {
      expect(mockNavigation.navigate).toHaveBeenCalledWith('TailorTech');
    }
  });

  it('redirects to TailorTech if user is already logged in', async () => {
    mockUserContext.user = { id: 1, name: 'John Doe', email: 'john.doe@example.com', phoneNumber: '', address: '', points: 0};
    (axios.get as jest.MockedFunction<any>).mockResolvedValue({ data: mockUserContext.user });

    await act(async () => {
      renderer.create(<LoginScreen {...mockProps} />);
    });

    expect(mockNavigation.navigate).toHaveBeenCalledWith('TailorTech');
  });
});
