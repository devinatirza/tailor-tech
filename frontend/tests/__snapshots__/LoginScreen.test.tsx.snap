// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen handles failed login 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#F8F8F8",
      "flex": 1,
      "justifyContent": "center",
      "paddingHorizontal": 40,
    }
  }
>
  <Text
    style={
      {
        "color": "#401201",
        "fontSize": 24,
        "fontWeight": "bold",
        "marginBottom": 20,
        "textAlign": "center",
      }
    }
  >
    Welcome Back!
  </Text>
  <View
    style={
      {
        "marginBottom": 20,
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "borderBottomColor": "#401201",
          "borderBottomWidth": 1,
          "marginBottom": 20,
          "width": "100%",
        }
      }
    >
      <TextInput
        onChangeText={[Function]}
        placeholder="Email"
        style={
          {
            "fontSize": 16,
            "height": 50,
            "paddingHorizontal": 10,
          }
        }
        value="test@example.com"
      />
    </View>
    <View
      style={
        {
          "borderBottomColor": "#401201",
          "borderBottomWidth": 1,
          "marginBottom": 20,
          "width": "100%",
        }
      }
    >
      <TextInput
        onChangeText={[Function]}
        placeholder="Password"
        secureTextEntry={true}
        style={
          {
            "fontSize": 16,
            "height": 50,
            "paddingHorizontal": 10,
          }
        }
        value="wrongpassword"
      />
    </View>
  </View>
  <Text
    style={
      {
        "color": "red",
        "marginBottom": 10,
      }
    }
  >
    Invalid credentials
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "opacity": 1,
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#D9C3A9",
          "borderRadius": 5,
          "height": 50,
          "justifyContent": "center",
          "marginBottom": 10,
          "width": "60%",
        }
      }
    >
      <Text
        style={
          {
            "color": "#260101",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Login
      </Text>
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 16,
          "marginBottom": 10,
          "textAlign": "center",
        }
      }
    >
      Didn't have an account?
    </Text>
  </View>
</View>
`;

exports[`LoginScreen handles successful login 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#F8F8F8",
      "flex": 1,
      "justifyContent": "center",
      "paddingHorizontal": 40,
    }
  }
>
  <Text
    style={
      {
        "color": "#401201",
        "fontSize": 24,
        "fontWeight": "bold",
        "marginBottom": 20,
        "textAlign": "center",
      }
    }
  >
    Welcome Back!
  </Text>
  <View
    style={
      {
        "marginBottom": 20,
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "borderBottomColor": "#401201",
          "borderBottomWidth": 1,
          "marginBottom": 20,
          "width": "100%",
        }
      }
    >
      <TextInput
        onChangeText={[Function]}
        placeholder="Email"
        style={
          {
            "fontSize": 16,
            "height": 50,
            "paddingHorizontal": 10,
          }
        }
        value=""
      />
    </View>
    <View
      style={
        {
          "borderBottomColor": "#401201",
          "borderBottomWidth": 1,
          "marginBottom": 20,
          "width": "100%",
        }
      }
    >
      <TextInput
        onChangeText={[Function]}
        placeholder="Password"
        secureTextEntry={true}
        style={
          {
            "fontSize": 16,
            "height": 50,
            "paddingHorizontal": 10,
          }
        }
        value=""
      />
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "opacity": 1,
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#D9C3A9",
          "borderRadius": 5,
          "height": 50,
          "justifyContent": "center",
          "marginBottom": 10,
          "width": "60%",
        }
      }
    >
      <Text
        style={
          {
            "color": "#260101",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Login
      </Text>
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 16,
          "marginBottom": 10,
          "textAlign": "center",
        }
      }
    >
      Didn't have an account?
    </Text>
  </View>
</View>
`;

exports[`LoginScreen navigates to Register screen 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#F8F8F8",
      "flex": 1,
      "justifyContent": "center",
      "paddingHorizontal": 40,
    }
  }
>
  <Text
    style={
      {
        "color": "#401201",
        "fontSize": 24,
        "fontWeight": "bold",
        "marginBottom": 20,
        "textAlign": "center",
      }
    }
  >
    Welcome Back!
  </Text>
  <View
    style={
      {
        "marginBottom": 20,
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "borderBottomColor": "#401201",
          "borderBottomWidth": 1,
          "marginBottom": 20,
          "width": "100%",
        }
      }
    >
      <TextInput
        onChangeText={[Function]}
        placeholder="Email"
        style={
          {
            "fontSize": 16,
            "height": 50,
            "paddingHorizontal": 10,
          }
        }
        value=""
      />
    </View>
    <View
      style={
        {
          "borderBottomColor": "#401201",
          "borderBottomWidth": 1,
          "marginBottom": 20,
          "width": "100%",
        }
      }
    >
      <TextInput
        onChangeText={[Function]}
        placeholder="Password"
        secureTextEntry={true}
        style={
          {
            "fontSize": 16,
            "height": 50,
            "paddingHorizontal": 10,
          }
        }
        value=""
      />
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "opacity": 1,
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#D9C3A9",
          "borderRadius": 5,
          "height": 50,
          "justifyContent": "center",
          "marginBottom": 10,
          "width": "60%",
        }
      }
    >
      <Text
        style={
          {
            "color": "#260101",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Login
      </Text>
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 16,
          "marginBottom": 10,
          "textAlign": "center",
        }
      }
    >
      Didn't have an account?
    </Text>
  </View>
</View>
`;

exports[`LoginScreen renders login screen correctly (snapshot) 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#F8F8F8",
      "flex": 1,
      "justifyContent": "center",
      "paddingHorizontal": 40,
    }
  }
>
  <Text
    style={
      {
        "color": "#401201",
        "fontSize": 24,
        "fontWeight": "bold",
        "marginBottom": 20,
        "textAlign": "center",
      }
    }
  >
    Welcome Back!
  </Text>
  <View
    style={
      {
        "marginBottom": 20,
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "borderBottomColor": "#401201",
          "borderBottomWidth": 1,
          "marginBottom": 20,
          "width": "100%",
        }
      }
    >
      <TextInput
        onChangeText={[Function]}
        placeholder="Email"
        style={
          {
            "fontSize": 16,
            "height": 50,
            "paddingHorizontal": 10,
          }
        }
        value=""
      />
    </View>
    <View
      style={
        {
          "borderBottomColor": "#401201",
          "borderBottomWidth": 1,
          "marginBottom": 20,
          "width": "100%",
        }
      }
    >
      <TextInput
        onChangeText={[Function]}
        placeholder="Password"
        secureTextEntry={true}
        style={
          {
            "fontSize": 16,
            "height": 50,
            "paddingHorizontal": 10,
          }
        }
        value=""
      />
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "opacity": 1,
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#D9C3A9",
          "borderRadius": 5,
          "height": 50,
          "justifyContent": "center",
          "marginBottom": 10,
          "width": "60%",
        }
      }
    >
      <Text
        style={
          {
            "color": "#260101",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Login
      </Text>
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 16,
          "marginBottom": 10,
          "textAlign": "center",
        }
      }
    >
      Didn't have an account?
    </Text>
  </View>
</View>
`;
